pkg/controller/ttlafterfinished/util.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ttlafterfinished
+
+import (
+	"time"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func NewJobsDelayCleanupQueue() *cache.Heap {
+	keyFn := cache.DeletionHandlingMetaNamespaceKeyFunc
+	lessFn := soonerExpiringJob
+	return cache.NewHeap(keyFn, lessFn)
+}
+
+// soonerExpiringJob is a cache.LessFunc used to compare expiring Jobs in a heap.
+func soonerExpiringJob(finishedJob1, finishedJob2 interface{}) bool {
+	j1, ok1 := finishedJob1.(*batch.Job)
+	j2, ok2 := finishedJob2.(*batch.Job)
+	// Error case: non-Job objects are considered 'smaller'.
+	// If they're both non-Job, they're considered equal.
+	if !ok1 && !ok2 {
+		return false
+	}
+	if !ok1 {
+		return true
+	}
+	if !ok2 {
+		return false
+	}
+	// Error case: Jobs without expiration time are considered 'smaller'.
```

Comment:
Jobs with no expiration time shouldn't stay in the delay clean up queue (as we won't clean them up), so pop them first instead of letting them occupy the space. Updated comments to clarify it.